common/log.h

```diff
@@ -47,6 +50,396 @@
 //
 // --------------------------------
 
+#ifdef LOG_WITH_TEST
+inline void log_test();
+#endif
+
+// Utility for synchronizing log configuration state
+//  since std::optional was introduced only in c++17
+enum class LogTriState
+{
+    Same,
+    False,
+    True
+};
+
+enum class LogTargetChannel
+{
+    Logfile,
+    Screen
+};
+
+class LogStateWrapper
+{
+    protected:
+        LogStateWrapper(){};
+
+        ~LogStateWrapper()
+        {
+            log_flush_all_targets();
+            for(auto t : _targets){ delete t; }
+        }
+
+    private:
+        LogStateWrapper(LogStateWrapper & other) = delete;
+        void operator=(const LogStateWrapper &)  = delete;
+
+    private:
+        static LogStateWrapper & instance()
+        {
+            static LogStateWrapper inst;
+            return inst;
+        }
```

Comment:
Singletons are generally considered an anti-pattern. Maybe a namespace, or possibly a class with a deleted constructor and static fields and methods, would be more appropriate.

Reply:
While I realize that most people seem to hate singletons with passion, in this particular case it's doing exactly what I wanted it to do, by being global and having runtime initialization, and without introducing additional "log_initialize" functionality, singleton or not, something will end up being a function, returning it's local static variable (struct/object), at least the way I'm thinking about it now.

If you have a particular problem in mind, that this can cause, I'll do it your way in an instant, no questions asked.

But otherwise, I'm leaning towards this staying this way, unless I figure a way to do it without making it more complicated.

I'll sleep on it, and see if I can figure out something better tomorrow.

Reply:
You could also just have a single, global instance of the class, and make any methods that call instance() non-static.

Reply:
Hmmm, I'm trying to visialize if this doesn't create a potential situation where a `LOG` might be called before that instance is initialized. I think I would have to ensure its initialized in main thread, before any other threads are created

Reply:
C++ has global constructors - you'll be fine as long as you initialize it on the same line as the declaration.

Reply:
Ok, I'll try tomorrow. I'm having a bit of flu/cold thing, I hope tomorrow my head will be a bit less mushy than right now :)

Reply:
Nothing brightens your day more than reading what isocpp has to say about something, only to find *"TODO: WRITE THIS UP"* :)

(https://isocpp.org/wiki/faq/ctors#nifty-counter-idiom)

Anyway, either I convert this class to fully static and have to rely on `atexit` for cleanup, or use global instance and constructor can't be private so single instance can't be guaranteed, or I have to use this shenaniganery: https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Nifty_Counter

What do you think ?

Reply:
If you really want to make sure there is only a single instance, you can use a global or 'static' variable to track whether it was initialized, and assert that it is false when you enter the constructor - then set it to true. I wouldn't worry about public vs private too much in a code base that mixes C and C++.

Reply:
Now that I think about it, if somebody can confirm or deny whether support for multiple targets ( log files ) has any use here, if I drop it, I could probably completely get rid of that entire singleton problem, altogether.